src/conversions.rs

```diff
@@ -30,77 +34,190 @@ impl Optimizer {
+                    Box::new(self.expr_to_structured_block(else_branch)),
+                ),
+                ("Loop", [block]) => {
+                    StructuredBlock::Loop(Box::new(self.expr_to_structured_block(block)))
+                }
+                ("Sequence", [block, rest]) => StructuredBlock::Sequence(vec![
+                    self.expr_to_structured_block(block),
+                    self.expr_to_structured_block(rest),
+                ]),
+                ("Break", [n]) => {
+                    if let Expr::Lit(egglog::ast::Literal::Int(n)) = n {
+                        StructuredBlock::Break(*n)
+                    } else {
+                        panic!("expected int literal for break");
+                    }
+                }
+                ("Return", [val]) => {
+                    if let Expr::Call(op, args) = val {
+                        match op.as_str() {
+                            "Void" => StructuredBlock::Return(None),
+                            "ReturnValue" => {
+                                assert_eq!(args.len(), 1);
+                                match &args[0] {
+                                    Expr::Lit(egglog::ast::Literal::String(val)) => {
+                                        StructuredBlock::Return(Some(val.to_string()))
+                                    }
+                                    _ => panic!("expected string literal for return value"),
+                                }
+                            }
+                            _ => panic!("expected void or return value"),
+                        }
+                    } else {
+                        panic!("expected call for return");
+                    }
+                }
+                _ => panic!("unknown structured block"),
+            }
+        } else {
+            panic!("expected call in expr_to_structured_block");
+        }
+    }
+
+    pub(crate) fn expr_to_basic_block(&mut self, expr: &Expr) -> BasicBlock {
+        if let Expr::Call(op, args) = expr {
+            assert_eq!(op.as_str(), "BlockNamed");
+
+            match &args.as_slice() {
+                [Expr::Lit(egglog::ast::Literal::String(name)), code] => {
+                    let code_vec = self.codelist_to_vec(code);
+                    let mut instrs = vec![];
+
+                    for expr in code_vec {
+                        self.expr_to_instructions(&expr, &mut instrs);
+                    }
+
+                    BasicBlock {
+                        name: BlockName::Named(name.to_string()),
+                        instrs,
+                        pos: None,
+                    }
+                }
+                _ => panic!("expected 2 args in expr_to_basic_block"),
+            }
+        } else {
+            panic!("expected call in expr_to_basic_block");
+        }
+    }
+
+    fn codelist_to_vec_helper(expr: &Expr, res: &mut Vec<Expr>) {
```

Comment:
is there a reason this is a void function with a mutable reference instead of returning the vec?

Reply:
the recursion would make this expensive- we could convert to linked list but it is slightly less performant